engine/crates/parser-sdl/src/rules/basic_type.rs

```diff
@@ -54,20 +67,56 @@ impl<'a> Visitor<'a> for BasicType {
                             resolver,
                             ..Default::default()
                         }
-                    })
+                    }),
                 )
-                    .with_description(type_definition.node.description.clone().map(|x| x.node))
-                    .with_cache_control(CacheDirective::parse(&type_definition.node.directives))
-                    .into()
-                ,&type_name,
-                &type_name
-                );
-
-                // If the type is a non primitive and also not modelized, it means we need to
-                // create the Input version of it.
-                // If the input is non used by other queries/mutation, it'll be removed from the
-                // final schema.
-                add_input_type_non_primitive(ctx, object, &type_name);
+                .with_description(type_definition.node.description.clone().map(|x| x.node))
+                .with_cache_control(CacheDirective::parse(&type_definition.node.directives))
+                .into()
+            },
+            &type_name,
+            &type_name,
+        );
+
+        // If the type is a non primitive and also not modelized, it means we need to
+        // create the Input version of it.
+        // If the input is non used by other queries/mutation, it'll be removed from the
+        // final schema.
+        add_input_type_non_primitive(ctx, object, &type_name);
+
+        // We also need to parse any unresolvable keys
+        let key_directives = directives
+            .iter()
+            .filter(|directive| directive.node.name.node == "key")
+            .collect::<Vec<_>>();
+
+        if !key_directives.is_empty() {
+            let (oks, errors) = key_directives
+                .into_iter()
+                .map(|directive| {
+                    Ok((
+                        directive.pos,
+                        parse_directive::<KeyDirective>(directive, ctx.variables)?,
+                    ))
+                })
+                .partition_result::<Vec<_>, Vec<_>, _, _>();
+
+            ctx.append_errors(errors);
+
+            for (pos, key) in oks {
+                if key.resolvable {
+                    ctx.report_error(
+                        vec![pos],
+                        "Found a resolvable key on a basic type, which is currently unsupported",
```

Comment:
Am I reading this file correctly, that BasicType is a GraphQL output type, in contrast to a model? (`type Foo { ... }`?)

I first read about resolvable and non-resolvable keys this afternoon so I'm likely missing important bits, but wouldn't we need to support both resolvable and unresolvable keys to define subgraphs? The basic thing I don't understand about this error is what is missing to make them work, in basic scenarios it wouldn't influence how the object is returned. (does it have to do with `__Entities`?)

Reply:
Note: I haven't had time to read the internal design docs on federation yet, I'm doing that next and the answer is probably there.